test/e2e/e2e_test.go

```diff
@@ -642,6 +642,68 @@ func checkMetricsCollection(t *testing.T, promClient prometheusv1.API, namespace
 	}
 }
 
+func runPromQuery(t *testing.T, promClient prometheusv1.API, query string, sampleTime time.Time) (float64, error) {
+	results, warnings, err := promClient.Query(context.Background(), query, sampleTime)
+	if err != nil {
+		t.Errorf("failed to execute prometheus query: %v", err)
+		return 0.0, err
+	}
+	if len(warnings) > 0 {
+		t.Logf("prometheus query emitted warnings: %v", warnings)
+	}
+
+	have, ok := results.(*model.Scalar)
+	if !ok {
+		t.Fatalf("expecting scalar type result")
+	}
+
+	return float64(have.Value), nil
+}
+
+func checkServiceCAMetrics(t *testing.T, client *kubernetes.Clientset, promClient prometheusv1.API) {
+	timeout := 60 * time.Second
+
+	secret, err := client.CoreV1().Secrets(serviceCAControllerNamespace).Get(signingKeySecretName, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("error retrieving signing key secret: %v", err)
+	}
+	currentCACerts, err := util.PemToCerts(secret.Data[v1.TLSCertKey])
+	if err != nil {
+		t.Fatalf("error unmarshaling %q: %v", v1.TLSCertKey, err)
+	}
+	if len(currentCACerts) == 0 {
+		t.Fatalf("no signing keys found")
+	}
+
+	sampleTime := time.Now()
+	want := currentCACerts[0].NotAfter
+	err = wait.PollImmediate(10*time.Second, timeout, func() (bool, error) {
+		have, err := runPromQuery(t, promClient, `service_ca_expiry_time`, sampleTime)
+		if err != nil {
+			t.Errorf("failed to execute prometheus query: %v", err)
+			return false, nil
+		}
+
+		if float64(want.Unix()) != have {
+			t.Fatalf("service ca expiry time mismatch expected %v observed %v", want, have)
+		}
+
+		have, err = runPromQuery(t, promClient, `service_ca_time_left`, sampleTime)
+		if err != nil {
+			t.Errorf("failed to execute prometheus query: %v", err)
+			return false, nil
+		}
+
+		if want.Sub(sampleTime).Seconds() != have {
+			t.Fatalf("time left for service ca expiry did not match. have %v want %v", have, want.Sub(sampleTime).Seconds())
+		}
+		return true, nil
+	})
+	if err != nil {
+		t.Fatalf("service ca expiry timer metrics collection failed %v", err)
```

Comment:
nit: Given that the only error that will appear here is due to timeout, maybe `("Failed to verify ca expiry metrics within %v", timeout)`?

nits:

- The first word of error messages should be lower-case because error messages are frequently aggregated in log output.
- The first word of a logged message (including to test logging) should instead be capitalized because it will be output directly rather than aggregated for output.
- Please ensure separation of message from serialized error (e.g. "<message>: %v" instead of "<message> %v").

Reply:
Using these code guidelines in recent commits